### PR TITLE
[KeyVault] BREAKING CHANGE: Support soft-delete feature for managed-HSM. `keyvault delete --hsm-name` will perform soft delete on a MHSM.

### PR DESCRIPTION
**Pull request by houk-ms**
**Description**

This PR supports the soft-delete feature for managed-HSM

The new added commands and parameters include
- `az keyvault show-deleted` command
- `--resource-type hsm` parameter for command `az keyvault list-deleted`
- `--hsm-name` parameter for command `az keyvault purge`

The following command's behavior changes
- `az keyvault delete --hsm-name `

It soft deletes a mhsm not permanently delete it. For permanently delete casse we provide the command `az keyvault purge --hsm-name`.

**Testing Guide**

**History Notes**

[Component Name 1] BREAKING CHANGE: az command a: Make some customer-facing breaking change.
[Component Name 2] az command b: Add some customer-facing feature.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

File: src/azure-cli-core/azure/cli/core/commands/__init__.py [hunk 1]
```diff
@@ -984,12 +985,13 @@ def __call__(self, poller):
 
         try:
             result = poller.result()
-        except ClientException as client_exception:
+        except (ClientException, HttpResponseError) as client_exception:
```
Comment (houk-ms) [inline src/azure-cli-core/azure/cli/core/commands/__init__.py:988]: We will handle the case in anothor PR https://github.com/Azure/azure-cli/pull/17833

File: src/azure-cli/azure/cli/command_modules/keyvault/_validators.py [hunk 2]
```diff
@@ -289,8 +289,8 @@ def validate_deleted_vault_or_hsm_name(cmd, ns):
     vault_name = getattr(ns, 'vault_name', None)
     hsm_name = getattr(ns, 'hsm_name', None)
 
-    if hsm_name:
-        raise InvalidArgumentValueError('Operation "purge" has not been supported for HSM.')
+    if hsm_name and 'keyvault recover' in cmd.name:
```
Comment (houk-ms) [inline src/azure-cli/azure/cli/command_modules/keyvault/_validators.py:292]: A bug fix here. Run `az keyvault recover` will show error messgae `purge has not been supported for HSM`

File: src/azure-cli/azure/cli/command_modules/keyvault/_validators.py [hunk 3]
```diff
@@ -312,25 +312,26 @@ def validate_deleted_vault_or_hsm_name(cmd, ns):
...
+                ns.location = resource.properties.location if vault_name \
+                    else resource.additional_properties.get('location')
                 break
 
     # if the vault was not found, throw an error
     if not resource:
         raise CLIError('No deleted Vault or HSM was found with name ' + resource_name)
 
-    if 'keyvault purge' not in cmd.name:
+    if 'keyvault purge' not in cmd.name and 'keyvault show-deleted' not in cmd.name:
```
Comment (houk-ms) [inline src/azure-cli/azure/cli/command_modules/keyvault/_validators.py:334]: these two comands won't need `resource-group` parameter

File: src/azure-cli/azure/cli/command_modules/keyvault/tests/latest/test_keyvault_commands.py [hunk 4]
```diff
@@ -181,58 +181,53 @@ def test_keyvault_private_endpoint_connection(self, resource_group):
 
 
 class KeyVaultHSMMgmtScenarioTest(ScenarioTest):
-    @record_only()
```
Comment (houk-ms) [inline src/azure-cli/azure/cli/command_modules/keyvault/tests/latest/test_keyvault_commands.py:184]: Make this test can be lively runned.

**Comment by yungezz** [comment]
this is not a CLI interface breaking, but service behavior change, right?

**Review by yungezz — APPROVED** [review]

**Comment by houk-ms** [comment]
> this is not a CLI interface breaking, but service behavior change, right?

Yes, this is a service side breaking change.

**Review by qwordy — APPROVED** [review]
Approve servicefabric

**Comment by qwordy** [comment]
> It soft deletes a mhsm not permanently delete it.

It is not a correct English sentence.

**Pull request merged**